### PR TITLE
verifier: forward tx to sequencer based on flag

### PR DESCRIPTION
**Pull request by tuxcanfly**
**Description**
This PR allows verifiers to optionally forward raw transactions to a sequencer by passing in a HTT URL pointing to the sequencer RPC endpoint.

**Additional context**
This flag is only used if the geth node is running in verifier mode.

**Metadata**
- Fixes #ENG-2214

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -1859,6 +1866,11 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *ethconfig.Config) {
 			cfg.EthDiscoveryURLs = SplitAndTrim(urls)
 		}
 	}
+	// Only configure sequencer http flag if we're running in verifier mode
+	// i.e. --mine is disabled.
```
Comment (protolambda) [inline cmd/utils/flags.go]:
```suggestion
	// Only configure sequencer http flag if we're running in verifier mode, i.e. --mine is disabled.
```

File: eth/api_backend.go [hunk 2]
```diff
@@ -244,6 +244,12 @@ func (b *EthAPIBackend) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscri
 }
 
 func (b *EthAPIBackend) SendTx(ctx context.Context, signedTx *types.Transaction) error {
+	if b.eth.txForwarder != nil {
+		err := b.eth.txForwarder.Forward(*signedTx)
+		if err != nil {
```
Comment (protolambda) [inline eth/api_backend.go]:
```suggestion
		if err := b.eth.txForwarder.Forward(signedTx); err != nil {
```

File: eth/tx_forwader.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	client, err := dialSequencerClientWithTimeout(f.ctx, f.SequencerHTTP)
+	if err != nil {
+		return err
+	}
+	f.Client = client
+	return nil
+}
+
+// Forward sends the transaction to the sequencer.
+func (f *TxForwarder) Forward(tx types.Transaction) error {
```
Comment (protolambda) [inline eth/tx_forwader.go]:
```suggestion
func (f *TxForwarder) Forward(tx *types.Transaction) error {
```

File: eth/tx_forwader.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	return ethclient.DialContext(ctxt, url)
+}
+
+// TxForwarder connects to a sequencer http endpoint and forwards mempool
+// transactions to it.
+type TxForwarder struct {
+	SequencerHTTP string
+	Client        *ethclient.Client
+	ctx           context.Context
```
Comment (protolambda) [inline eth/tx_forwader.go]:
```suggestion
```

File: eth/tx_forwader.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// Dial connects to the sequencer http endpoint. It only needs to be called
+// once after initializing a new forwarder.
+func (f *TxForwarder) Dial() error {
+	client, err := dialSequencerClientWithTimeout(f.ctx, f.SequencerHTTP)
```
Comment (protolambda) [inline eth/tx_forwader.go]:
```suggestion
func (f *TxForwarder) Dial(ctx context.Context) error {
	client, err := dialSequencerClientWithTimeout(ctx, f.SequencerHTTP)
```

File: eth/tx_forwader.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	return nil
+}
+
+// NewTxForwarder returns a new forwarder.
+func NewTxForwarder(rpc string) *TxForwarder {
+	return &TxForwarder{
+		ctx:           context.Background(),
+		SequencerHTTP: rpc,
+	}
+}
```
Comment (protolambda) [inline eth/tx_forwader.go]: The fields are all public, no need for a special New function. Saves us some diff lines too.

```suggestion
```

File: eth/tx_forwader.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	defer cancel()
+
+	return ethclient.DialContext(ctxt, url)
+}
+
+// TxForwarder connects to a sequencer http endpoint and forwards mempool
+// transactions to it.
+type TxForwarder struct {
+	SequencerHTTP string
+	Client        *ethclient.Client
```
Comment (protolambda) [inline eth/tx_forwader.go]:
```suggestion
	client        *ethclient.Client
```
Don't need to export this field

File: eth/tx_forwader.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// Dial connects to the sequencer http endpoint. It only needs to be called
+// once after initializing a new forwarder.
+func (f *TxForwarder) Dial() error {
+	client, err := dialSequencerClientWithTimeout(f.ctx, f.SequencerHTTP)
+	if err != nil {
+		return err
+	}
+	f.Client = client
```
Comment (protolambda) [inline eth/tx_forwader.go]: If we retain the client, should we close it?

File: eth/tx_forwader.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+// provided URL. If the dial doesn't complete within defaultDialTimeout
+// seconds, this method will return an error.
+func dialSequencerClientWithTimeout(ctx context.Context, url string) (
+	*ethclient.Client, error) {
+
+	ctxt, cancel := context.WithTimeout(ctx, defaultDialTimeout)
+	defer cancel()
+
+	return ethclient.DialContext(ctxt, url)
+}
```
Comment (protolambda) [inline eth/tx_forwader.go]: If we're only doing the dial in the `TxForwarder` I like to create the timeout context and call the ethclient dial there directly. The verbosity of this function is not worth it: we're trying to minimize the differences with geth, and so we should at most ever have one dial in this package like this.

File: eth/backend.go [hunk 10]
```diff
@@ -220,6 +221,10 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
 	}
 	eth.txPool = core.NewTxPool(config.TxPool, chainConfig, eth.blockchain)
 
+	if config.RollupSequencerHTTP != "" {
+		eth.txForwarder = NewTxForwarder(config.RollupSequencerHTTP)
```
Comment (protolambda) [inline eth/backend.go]:
```suggestion
		eth.txForwarder = &TxForwarder{SequencerHTTP: config.RollupSequencerHTTP}
```

File: eth/tx_forwader.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	f.Client = client
+	return nil
+}
+
+// Forward sends the transaction to the sequencer.
+func (f *TxForwarder) Forward(tx types.Transaction) error {
+	if f.Client == nil {
+		return errors.New("cannot forward without Dial")
+	}
+	return nil
```
Comment (protolambda) [inline eth/tx_forwader.go]: It's not forwarding yet? Should there be an ethclient call here?

**Review by protolambda — COMMENTED** [review]
Some review suggestions how I think the code can improve. But TBH I'd like to remove the `tx_forwarder` entirely: we can just use a `*ethclient.Client` type directly. It'll be less verbose, and in the future we may want to forward more than just sending a tx. E.g. sending a raw tx (there are multiple RPC endpoints).

File: eth/tx_forwader.go (hunk 11 above)
Comment (tuxcanfly) [reply eth/tx_forwader.go]: Yeah, sorry missed that.

**Comment by tuxcanfly** [comment]
Refactored to use RPC client to avoid import cycle, also removed tx forwarder struct and updated to instead just rely on the RPC client directly.

**Review by trianglesphere — APPROVED** [review]

**Comment by trianglesphere** [comment]
please squash this when merging. Also the base branch needs to be changed to `optimism-history`

**Comment by tuxcanfly** [comment]
Squashed and changed the base branch to `optimism-history`.

**Review by protolambda — APPROVED** [review]

**Pull request merged**